File: 2.add-two-numbers.js
/*
 * @lc app=leetcode id=2 lang=javascript
 *
 * [2] Add Two Numbers
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function (l1, l2, plus = 0) {
    if (l1 && l2) {
        return new ListNode((l1.val + l2.val + plus) % 10, addTwoNumbers(l1.next, l2.next, ~~((l1.val + l2.val + plus) / 10)))
    } else if (l1 || l2) {
        return new ListNode(((l1 ? l1.val : l2.val) + plus) % 10, addTwoNumbers(l1 ? l1.next : l2.next, null, ~~(((l1 ? l1.val : l2.val) + plus) / 10)));
    } else {
        if (plus)
            return new ListNode(1);
        else
            return null;
    }
};
// @lc code=end

File: 2.add-two-numbers.py
#
# @lc app=leetcode id=2 lang=python3
#
# [2] Add Two Numbers
#

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        head = ListNode(0)
        cur = head
        plus = 0
        while l1 or l2 or plus == 1:
            v1 = l1.val if l1 else 0
            v2 = l2.val if l2 else 0
            total = v1 + v2 + plus
            plus = total // 10

            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None

            cur.next = ListNode(total % 10)
            cur = cur.next

        return head.next
# @lc code=end
